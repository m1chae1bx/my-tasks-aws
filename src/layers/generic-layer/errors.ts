export interface CustomError extends Error {
  errorCode: string;
}

export class UsernameUnavailableError implements CustomError {
  name = "UsernameUnavailableError";
  message = "Username is already taken";
  errorCode = "usernameUnavailable";
}

export class EmailUnavailableError implements CustomError {
  name = "EmailUnavailableError";
  message = "Email address is already registered";
  errorCode = "emailUnavailable";
}
