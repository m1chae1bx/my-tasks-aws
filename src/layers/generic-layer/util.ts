import { APIGatewayProxyResult } from "aws-lambda";
import { AWSError } from "aws-sdk";
import addFormats from "ajv-formats";

export { v4 as uuid } from "uuid";
import { v4 as uuid } from "uuid";
export { default as Ajv, JSONSchemaType } from "ajv";
import Ajv from "ajv";

const ajv = new Ajv({ allErrors: true, $data: true });
addFormats(ajv);
export { ajv };

export const isAWSError = (error: unknown): error is AWSError => {
  return (error as AWSError).code !== undefined;
};

export const genericErrorHandler = (
  error: unknown,
  message: string,
  code?: string,
  context?: unknown
): APIGatewayProxyResult => {
  const referenceId = uuid();
  console.error({ error, context, referenceId });
  return {
    statusCode: 500,
    body: JSON.stringify({
      message,
      code,
      referenceId,
    }),
  };
};
