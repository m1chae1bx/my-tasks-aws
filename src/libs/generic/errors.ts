export interface CustomError extends Error {
  errorCode: ErrorCode;
}

export enum ErrorCode {
  USERNAME_UNAVAILABLE = "usernameUnavailable",
  USER_NOT_FOUND = "userNotFound",
  EMAIL_UNAVAILABLE = "emailUnavailable",
  REQUIRED_PROPERTY_MISSING = "requiredPropertyMissing",
  ENVIRONMENT_CONFIG_ERROR = "environmentConfigError",
  ID_ALREADY_EXISTS = "idAlreadyExists",
  TASK_NOT_FOUND_ERROR = "taskNotFound",
  LIST_NOT_FOUND = "listNotFound",
}

export class UsernameUnavailableError implements CustomError {
  name = "UsernameUnavailableError";
  message = "Username is already taken";
  errorCode = ErrorCode.USERNAME_UNAVAILABLE;
}

export class EmailUnavailableError implements CustomError {
  name = "EmailUnavailableError";
  message = "Email address is already registered";
  errorCode = ErrorCode.EMAIL_UNAVAILABLE;
}

export class UserNotFoundError implements CustomError {
  name = "UserNotFoundError";
  message = "User not found";
  errorCode = ErrorCode.USER_NOT_FOUND;
}

export class EnvironmentConfigError implements CustomError {
  name = "EnvironmentConfigError";
  message;
  errorCode = ErrorCode.ENVIRONMENT_CONFIG_ERROR;

  constructor(name: string) {
    this.message = `Environment config ${name} is not set`;
  }
}

export class RequiredPropertyMissingError implements CustomError {
  name = "RequiredPropertyMissingError";
  message;
  errorCode = ErrorCode.REQUIRED_PROPERTY_MISSING;

  constructor(name: string) {
    this.message = `Property ${name} is required`;
  }
}

export class IdAlreadyExistsError implements CustomError {
  name = "IdAlreadyExistsError";
  message;
  errorCode = ErrorCode.ID_ALREADY_EXISTS;

  constructor(entity: string) {
    this.message = `${entity} ID already exists`;
  }
}

export class TaskNotFoundError implements CustomError {
  name = "TaskNotFoundError";
  message;
  errorCode = ErrorCode.TASK_NOT_FOUND_ERROR;

  constructor(id: string) {
    this.message = `Task with ID ${id} not found`;
  }
}

export class ListNotFoundError implements CustomError {
  name = "ListNotFoundError";
  message;
  errorCode = ErrorCode.LIST_NOT_FOUND;

  constructor(id: string) {
    this.message = `List with ID ${id} not found`;
  }
}
