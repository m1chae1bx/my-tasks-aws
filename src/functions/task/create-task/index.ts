import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : {},
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { name, dueDate, desc } = request.body;
    const { listId } = request.pathParameters;
    const task = new Task(
      listId,
      name,
      false,
      dueDate ? new Date(dueDate) : undefined,
      desc
    );

    const id = await task.save();
    const response = { message: `Task '${name}' was created successfully`, id };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while creating the task. Please try again later."
    );
  }
};
