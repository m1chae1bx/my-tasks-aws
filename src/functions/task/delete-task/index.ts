import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "@libs/task";
import { genericErrorHandler } from "@libs/generic/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { listId, taskId } = request.pathParameters;
    await Task.delete(taskId, listId);

    const response = { message: `Task ${taskId} was deleted successfully` };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while deleting the task. Please try again later.`
    );
  }
};
