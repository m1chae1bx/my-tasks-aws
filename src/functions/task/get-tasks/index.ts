import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      queryStringParameters: event.queryStringParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { listId } = request.pathParameters;
    const tasks = await Task.getAll(listId, request.queryStringParameters);
    return {
      statusCode: 200,
      body: JSON.stringify(tasks),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while getting the tasks. Please try again later."
    );
  }
};
