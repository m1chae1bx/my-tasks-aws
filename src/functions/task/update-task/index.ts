import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "@libs/task";
import { genericErrorHandler } from "@libs/generic/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : null,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const listId = request.pathParameters.listId;
    const { id, name, isCompleted, dueDate, desc } = request.body;

    const task = new Task(listId, name, isCompleted, dueDate, desc, id);

    await task.update();
    return {
      statusCode: 200,
      body: JSON.stringify({
        message: `Task '${name}' was updated successfully`,
      }),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while updating the task. Please try again later.`
    );
  }
};
