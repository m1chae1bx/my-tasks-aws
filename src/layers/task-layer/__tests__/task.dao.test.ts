import {
  completedTask,
  getTasksQuery,
  testTask,
  testTasks,
} from "../test-data";
import { dynamoClient } from "/opt/nodejs/dynamo.config";
import { CustomError, ErrorCode } from "/opt/nodejs/errors";
import * as TaskDao from "../task.dao";
import { isAWSError } from "/opt/nodejs/util";

jest.mock("/opt/nodejs/dynamo.config", () => ({
  dynamoClient: {
    put: jest.fn(),
    delete: jest.fn(),
    query: jest.fn(),
    transactWrite: jest.fn(),
  },
  getTableName: jest.fn().mockReturnValue("test-table-name"),
}));

jest.mock("/opt/nodejs/util", () => ({
  ...jest.requireActual("/opt/nodejs/util"),
  isAWSError: jest.fn(),
  uuid: jest.fn().mockReturnValue("test-uuid"),
}));

describe("create", () => {
  describe("happy path", () => {
    it("should return the task ID", async () => {
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce(null),
      });
      const result = await TaskDao.create(testTask);
      expect(result).toBe("test-uuid");
    });
  });

  describe("sad path - ID already exists", () => {
    let result: unknown;

    beforeAll(async () => {
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce({
          code: "ConditionalCheckFailedException",
        }),
      });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      jest
        .spyOn(console, "error")
        .mockClear()
        .mockImplementationOnce(jest.fn());
      try {
        await TaskDao.create(testTask);
      } catch (error) {
        result = error;
      }
    });

    it("should throw an error", () => {
      expect((result as CustomError).errorCode).toBe(
        ErrorCode.ID_ALREADY_EXISTS
      );
    });

    it("should log an error", async () => {
      expect(jest.spyOn(console, "error").mock.calls).toMatchSnapshot();
    });
  });

  describe("sad path - DynamoDB put unknown error", () => {
    it("should throw an error", async () => {
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce(new Error("test-error")),
      });

      try {
        await TaskDao.create(testTask);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });
});

describe("getAll", () => {
  describe("happy path", () => {
    it("should return a list of tasks", async () => {
      (dynamoClient.query as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce({
          Items: testTasks,
        }),
      });
      const result = await TaskDao.getAll("test-list-id", getTasksQuery);
      expect(result).toEqual(testTasks);
    });
  });

  describe("happy path - no items", () => {
    it("should return an empty list", async () => {
      (dynamoClient.query as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce({
          Items: [],
        }),
      });
      const result = await TaskDao.getAll("test-list-id", getTasksQuery);
      expect(result).toEqual([]);
    });
  });
});

describe("update", () => {
  describe("happy path - task completed", () => {
    it("should call DynamoDB transactWrite", async () => {
      jest.spyOn(dynamoClient, "transactWrite").mockClear();
      (dynamoClient.transactWrite as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce(null),
      });
      await TaskDao.update(completedTask);
      expect(
        jest.spyOn(dynamoClient, "transactWrite").mock.calls
      ).toMatchSnapshot();
    });
  });

  describe("happy path - active task", () => {
    it("should call DynamoDB put", async () => {
      jest.spyOn(dynamoClient, "put").mockClear();
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce(null),
      });
      await TaskDao.update(testTask);
      expect(jest.spyOn(dynamoClient, "put").mock.calls).toMatchSnapshot();
    });
  });

  describe("happy path - task reopened", () => {
    it("should call DynamoDB transactWrite", async () => {
      jest.spyOn(dynamoClient, "transactWrite").mockClear();
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce({
          code: "ConditionalCheckFailedException",
        }),
      });
      (dynamoClient.transactWrite as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce(null),
      });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      await TaskDao.update(testTask);
      expect(
        jest.spyOn(dynamoClient, "transactWrite").mock.calls
      ).toMatchSnapshot();
    });
  });

  describe("sad path - missing id", () => {
    it("should throw an error", async () => {
      try {
        await TaskDao.update({ ...testTask, id: undefined });
      } catch (error) {
        expect((error as CustomError).errorCode).toBe(
          ErrorCode.REQUIRED_PROPERTY_MISSING
        );
      }
    });
  });

  describe("sad path - failed updating active task", () => {
    it("should throw an error", async () => {
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce(new Error("test error")),
      });
      try {
        await TaskDao.update(testTask);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  describe("sad path - failed updating reopened task", () => {
    it("should throw an error", async () => {
      (dynamoClient.put as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce({
          code: "ConditionalCheckFailedException",
        }),
      });
      (dynamoClient.transactWrite as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce(new Error("test error")),
      });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      try {
        await TaskDao.update(testTask);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });
});

describe("deleteTask", () => {
  describe("happy path - delete active task", () => {
    it("should call DynamoDB delete operation", async () => {
      jest.spyOn(dynamoClient, "delete").mockClear();
      (dynamoClient.delete as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockResolvedValueOnce(null),
      });
      await TaskDao.deleteTask("test-task-id", "test-list-id");
      expect(jest.spyOn(dynamoClient, "delete").mock.calls).toMatchSnapshot();
    });
  });

  describe("happy path - delete completed task", () => {
    it("should call DynamoDB delete operation", async () => {
      jest.spyOn(dynamoClient, "delete").mockClear();
      (dynamoClient.delete as jest.Mock)
        .mockReturnValueOnce({
          promise: jest.fn().mockRejectedValueOnce({
            code: "ConditionalCheckFailedException",
          }),
        })
        .mockReturnValueOnce({
          promise: jest.fn().mockResolvedValueOnce(null),
        });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      await TaskDao.deleteTask("test-task-id", "test-list-id");
      expect(jest.spyOn(dynamoClient, "delete").mock.calls).toMatchSnapshot();
    });
  });

  describe("delete active task unknown error", () => {
    it("should throw an error", async () => {
      (dynamoClient.delete as jest.Mock).mockReturnValueOnce({
        promise: jest.fn().mockRejectedValueOnce(new Error("test error")),
      });
      try {
        await TaskDao.deleteTask("test-task-id", "test-list-id");
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  describe("delete completed task unknown error", () => {
    it("should throw an error", async () => {
      (dynamoClient.delete as jest.Mock)
        .mockReturnValueOnce({
          promise: jest.fn().mockRejectedValueOnce({
            code: "ConditionalCheckFailedException",
          }),
        })
        .mockReturnValueOnce({
          promise: jest.fn().mockRejectedValueOnce(new Error("test error")),
        });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      try {
        await TaskDao.deleteTask("test-task-id", "test-list-id");
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  describe("sad path - task not found error", () => {
    it("should throw an error", async () => {
      jest
        .spyOn(console, "error")
        .mockClear()
        .mockImplementationOnce(jest.fn());
      (dynamoClient.delete as jest.Mock)
        .mockReturnValueOnce({
          promise: jest.fn().mockRejectedValueOnce({
            code: "ConditionalCheckFailedException",
          }),
        })
        .mockReturnValueOnce({
          promise: jest.fn().mockRejectedValueOnce({
            code: "ConditionalCheckFailedException",
          }),
        });
      (isAWSError as unknown as jest.Mock).mockReturnValue(true);
      try {
        await TaskDao.deleteTask("test-task-id", "test-list-id");
      } catch (error) {
        expect((error as CustomError).errorCode).toBe(
          ErrorCode.TASK_NOT_FOUND_ERROR
        );
      }
    });
  });
});
