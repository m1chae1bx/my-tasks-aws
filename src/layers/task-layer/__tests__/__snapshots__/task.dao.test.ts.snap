// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create sad path - DynamoDB put unknown error should throw an error 1`] = `[Error: test-error]`;

exports[`create sad path - ID already exists should log an error 1`] = `
Array [
  Array [
    Object {
      "code": "ConditionalCheckFailedException",
    },
  ],
]
`;

exports[`deleteTask delete active task unknown error should throw an error 1`] = `[Error: test error]`;

exports[`deleteTask delete completed task unknown error should throw an error 1`] = `[Error: test error]`;

exports[`deleteTask happy path - delete active task should call DynamoDB delete operation 1`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_exists(PK)",
      "Key": Object {
        "PK": "LIST#test-list-id",
        "SK": "TASK#active#test-task-id",
      },
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`deleteTask happy path - delete completed task should call DynamoDB delete operation 1`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_exists(PK)",
      "Key": Object {
        "PK": "LIST#test-list-id",
        "SK": "TASK#test-task-id",
      },
      "TableName": "test-table-name",
    },
  ],
  Array [
    Object {
      "ConditionExpression": "attribute_exists(PK)",
      "Key": Object {
        "PK": "LIST#test-list-id",
        "SK": "TASK#test-task-id",
      },
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`getAll happy path - due date overdue should call DynamoDB query 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#SK": "SK",
        "#desc": "desc",
        "#name": "name",
      },
      "ExpressionAttributeValues": Object {
        ":PK": "LIST#test-list-id",
        ":SK": "TASK#",
        ":dueDate": "2020-01-01T00:00:00.000Z",
        ":name": "test task",
      },
      "FilterExpression": "contains(nameSearch, :name) and dueDate < :dueDate",
      "KeyConditionExpression": "PK = :PK and begins_with(#SK, :SK)",
      "ProjectionExpression": "id, #name, #desc, isCompleted, dueDate",
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`getAll happy path - due date tomorrow should call DynamoDB query 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#SK": "SK",
        "#desc": "desc",
        "#name": "name",
      },
      "ExpressionAttributeValues": Object {
        ":PK": "LIST#test-list-id",
        ":SK": "TASK#",
        ":dueDate": "2020-01-02T00:00:00.000Z",
        ":name": "test task",
      },
      "FilterExpression": "contains(nameSearch, :name) and dueDate = :dueDate",
      "KeyConditionExpression": "PK = :PK and begins_with(#SK, :SK)",
      "ProjectionExpression": "id, #name, #desc, isCompleted, dueDate",
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`getAll happy path - due date unplanned should call DynamoDB query 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#SK": "SK",
        "#desc": "desc",
        "#name": "name",
      },
      "ExpressionAttributeValues": Object {
        ":PK": "LIST#test-list-id",
        ":SK": "TASK#",
        ":name": "test task",
      },
      "FilterExpression": "contains(nameSearch, :name) and attribute_not_exists(dueDate)",
      "KeyConditionExpression": "PK = :PK and begins_with(#SK, :SK)",
      "ProjectionExpression": "id, #name, #desc, isCompleted, dueDate",
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`getAll happy path - due date upcoming should call DynamoDB query 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeNames": Object {
        "#SK": "SK",
        "#desc": "desc",
        "#name": "name",
      },
      "ExpressionAttributeValues": Object {
        ":PK": "LIST#test-list-id",
        ":SK": "TASK#",
        ":dueDate": "2020-01-02T00:00:00.000Z",
        ":name": "test task",
      },
      "FilterExpression": "contains(nameSearch, :name) and dueDate > :dueDate",
      "KeyConditionExpression": "PK = :PK and begins_with(#SK, :SK)",
      "ProjectionExpression": "id, #name, #desc, isCompleted, dueDate",
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`update happy path - active task should call DynamoDB put 1`] = `
Array [
  Array [
    Object {
      "ConditionExpression": "attribute_exists(PK)",
      "Item": Object {
        "PK": "LIST#test-list-id",
        "SK": "TASK#active#test-task-id",
        "desc": "Test task description",
        "dueDate": 2022-01-01T00:00:00.000Z,
        "id": "test-task-id",
        "isCompleted": false,
        "listId": "test-list-id",
        "name": "Test task",
      },
      "TableName": "test-table-name",
    },
  ],
]
`;

exports[`update happy path - task completed should call DynamoDB transactWrite 1`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Delete": Object {
            "Key": Object {
              "PK": "LIST#test-list-id",
              "SK": "TASK#active#test-task-id",
            },
            "TableName": "test-table-name",
          },
        },
        Object {
          "Put": Object {
            "Item": Object {
              "PK": "LIST#test-list-id",
              "SK": "TASK#test-task-id",
              "desc": "Test task description",
              "dueDate": 2022-01-01T00:00:00.000Z,
              "id": "test-task-id",
              "isCompleted": true,
              "listId": "test-list-id",
              "name": "Test task",
              "nameSearch": "test task",
            },
            "TableName": "test-table-name",
          },
        },
      ],
    },
  ],
]
`;

exports[`update happy path - task reopened should call DynamoDB transactWrite 1`] = `
Array [
  Array [
    Object {
      "TransactItems": Array [
        Object {
          "Delete": Object {
            "Key": Object {
              "PK": "LIST#test-list-id",
              "SK": "TASK#test-task-id",
            },
            "TableName": "test-table-name",
          },
        },
        Object {
          "Put": Object {
            "Item": Object {
              "PK": "LIST#test-list-id",
              "SK": "TASK#active#test-task-id",
              "desc": "Test task description",
              "dueDate": 2022-01-01T00:00:00.000Z,
              "id": "test-task-id",
              "isCompleted": false,
              "listId": "test-list-id",
              "name": "Test task",
            },
            "TableName": "test-table-name",
          },
        },
      ],
    },
  ],
]
`;

exports[`update sad path - failed updating active task should throw an error 1`] = `[Error: test error]`;

exports[`update sad path - failed updating reopened task should throw an error 1`] = `[Error: test error]`;
